taichi/backends/opengl/codegen_opengl.cpp

```diff
@@ -529,8 +529,47 @@ class KernelGen : public IRVisitor {
     const auto *argload = stmt->base_ptrs[0]->as<ArgLoadStmt>();
     const int arg_id = argload->arg_id;
     const int num_indices = stmt->indices.size();
+    auto element_shape = stmt->element_shape;
     std::vector<std::string> size_var_names;
-    for (int i = 0; i < num_indices; i++) {
+    std::vector<std::string> element_shape_size_var_names;
+    const int layout_SOA = 1;
+    const int layout_AOS = 0;
+    int layout = stmt->element_dim <= 0 ? layout_AOS : layout_SOA;
+
+    if (element_shape.size() > 0) {
+      int elem_beg = num_indices - element_shape.size();
+      int elem_end = num_indices;
+      if (layout == layout_SOA) {
+        elem_beg = 0;
+        elem_end = element_shape.size();
+      }
+      for (int i = elem_beg; i < elem_end; i++) {
+        used.int32 = true;
+        std::string var_name = fmt::format("_s{}_{}{}", i, "arr", arg_id);
+        if (!loaded_args_.count(var_name)) {
+          // The passed element shape vector is reversed w.r.t. to args buffer
+          emit("int {} = {};", var_name, element_shape[i - elem_beg]);
+          loaded_args_.insert(var_name);
+        }
+        element_shape_size_var_names.push_back(std::move(var_name));
+      }
+    }
+    // Args buffer arrange dimensions higher to lower
+    // AoS args buffer:   array_shape|element_shape
+    // SoA args buffer: element_shape|array_shape
+    //
+    // ti.Matrix.ndarray(3, 2, ti.f32, (5, 4), layout=ti.Layout.AOS)
+    // args buffer: 5, 4, 3, 2
+    // ti.Matrix.ndarray(3, 2, ti.f32, (5, 4), layout=ti.Layout.SOA)
+    // args buffer: 3, 2, 5, 4
+
+    int ind_beg = 0;
```

Comment:
To make it easier to read, can we set elem_beg/elem_end/ind_beg/ind_end in one if/else block? ;)

Reply:
Nice suggestion, thanks!